<H1>Importing Libraries</H1>

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Create path
path = r'C:\Users\supri\Desktop\Supriya\DataAnalysis\Project\Data Immersion\Achievement 1-2\Instacart_Basket_Analysis'


In [3]:
# Import dataframe including newly derived columns from the previous exercise
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_derived_new_variables.pkl'))

In [4]:
# Check output
ords_prods_merge.shape

(32404859, 23)

<H3>2.Find the aggregated mean of the “order_number” column grouped by “department_id”  for the entire dataframe</H3>

In [5]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

<H4>3 Observation - Compared to the subset, the full dataframe includes all 'department_id' values. Also, the mean values for department IDs found in the subset have changed—some increased, while others decreased. For instance, department ID 4 had a mean of 18.825780 in the subset, but in the full dataset, it is 17.811403.</H4>

<H3>4.Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.</H3>

In [7]:
# These operations were already performed on the dataframe during the previous exercise
#Can be confirmed by the presence of the 'max_orders' and 'loyalty_flag' columns already in place

In [8]:
ords_prods_merge.head(20)

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0              0   2539329        1    prior             1   
1              0   2539329        1    prior             1   
2              0   2539329        1    prior             1   
3              0   2539329        1    prior             1   
4              0   2539329        1    prior             1   
5              1   2398795        1    prior             2   
6              1   2398795        1    prior             2   
7              1   2398795        1    prior             2   
8              1   2398795        1    prior             2   
9              1   2398795        1    prior             2   
10             1   2398795        1    prior             2   
11             2    473747        1    prior             3   
12             2    473747        1    prior             3   
13             2    473747        1    prior             3   
14             2    473747        1    prior             3   
15             2    473747        1    prior             3   
16             3   2254736        1    prior             4   
17             3   2254736        1    prior             4   
18             3   2254736        1    prior             4   
19             3   2254736        1    prior             4   

    orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                     2                  8                     NaN   
1                     2                  8                     NaN   
2                     2                  8                     NaN   
3                     2                  8                     NaN   
4                     2                  8                     NaN   
5                     3                  7                    15.0   
6                     3                  7                    15.0   
7                     3                  7                    15.0   
8                     3                  7                    15.0   
9                     3                  7                    15.0   
10                    3                  7                    15.0   
11                    3                 12                    21.0   
12                    3                 12                    21.0   
13                    3                 12                    21.0   
14                    3                 12                    21.0   
15                    3                 12                    21.0   
16                    4                  7                    29.0   
17                    4                  7                    29.0   
18                    4                  7                    29.0   
19                    4                  7                    29.0   

    first_time_customers  product_id  ...  aisle_id  department_id prices  \
0                   True         196  ...        77              7    9.0   
1                   True       14084  ...        91             16   12.5   
2                   True       12427  ...        23             19    4.4   
3                   True       26088  ...        23             19    4.7   
4                   True       26405  ...        54             17    1.0   
5                  False         196  ...        77              7    9.0   
6                  False       10258  ...       117             19    3.0   
7                  False       12427  ...        23             19    4.4   
8                  False       13176  ...        24              4   10.3   
9                  False       26088  ...        23             19    4.7   
10                 False       13032  ...       121             14    4.0   
11                 False         196  ...        77              7    9.0   
12                 False       12427  ...        23             19    4.4   
13                 False       10258  ...       117             19    3.0   
14                 False       25133  ...        21             16    8.6  

In [9]:
# Check loyalty flags frequency
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

<H3>05. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.</H3>

In [10]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'sum']})

prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2

<H4>Observation -On average, new customers spend the most money on their purchases, followed by regular customers, and lastly loyal customers.</H4>

<H3>06. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”</H3>

In [11]:
# Create a new column for the average price of a product
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [12]:
# Check the output
ords_prods_merge.head(20)

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0              0   2539329        1    prior             1   
1              0   2539329        1    prior             1   
2              0   2539329        1    prior             1   
3              0   2539329        1    prior             1   
4              0   2539329        1    prior             1   
5              1   2398795        1    prior             2   
6              1   2398795        1    prior             2   
7              1   2398795        1    prior             2   
8              1   2398795        1    prior             2   
9              1   2398795        1    prior             2   
10             1   2398795        1    prior             2   
11             2    473747        1    prior             3   
12             2    473747        1    prior             3   
13             2    473747        1    prior             3   
14             2    473747        1    prior             3   
15             2    473747        1    prior             3   
16             3   2254736        1    prior             4   
17             3   2254736        1    prior             4   
18             3   2254736        1    prior             4   
19             3   2254736        1    prior             4   

    orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                     2                  8                     NaN   
1                     2                  8                     NaN   
2                     2                  8                     NaN   
3                     2                  8                     NaN   
4                     2                  8                     NaN   
5                     3                  7                    15.0   
6                     3                  7                    15.0   
7                     3                  7                    15.0   
8                     3                  7                    15.0   
9                     3                  7                    15.0   
10                    3                  7                    15.0   
11                    3                 12                    21.0   
12                    3                 12                    21.0   
13                    3                 12                    21.0   
14                    3                 12                    21.0   
15                    3                 12                    21.0   
16                    4                  7                    29.0   
17                    4                  7                    29.0   
18                    4                  7                    29.0   
19                    4                  7                    29.0   

    first_time_customers  product_id  ...  department_id  prices match  \
0                   True         196  ...              7     9.0  both   
1                   True       14084  ...             16    12.5  both   
2                   True       12427  ...             19     4.4  both   
3                   True       26088  ...             19     4.7  both   
4                   True       26405  ...             17     1.0  both   
5                  False         196  ...              7     9.0  both   
6                  False       10258  ...             19     3.0  both   
7                  False       12427  ...             19     4.4  both   
8                  False       13176  ...              4    10.3  both   
9                  False       26088  ...             19     4.7  both   
10                 False       13032  ...             14     4.0  both   
11                 False         196  ...              7     9.0  both   
12                 False       12427  ...             19     4.4  both   
13                 False       10258  ...             19     3.0  both   
14                 False       25133  ...             16     8.6  both   
15                 False       30450  ...     

In [13]:
# Setting the spending_flag with the criteria
ords_prods_merge.loc[ords_prods_merge['average_price']<10,'spending_flag']='Low_spender'

In [14]:
ords_prods_merge.loc[ords_prods_merge['average_price']>= 10, 'spending_flag'] = 'High_spender'

In [15]:
# Check the frequency of the spending-flag
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

<H4>Observation - Low numbers of high spenders compared to the Low spenders</H4>

<H3>07. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a 'Non-frequent customer'
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a 'Regular customer'
If the median is lower than or equal to 10, then the customer should be labeled a 'Frequent customer'</H3>

In [16]:
# Create a new column for customer frequency flags based on the median values from 'days_since_prior_order' column
ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [17]:
# Check output
ords_prods_merge.head(20)

Unnamed: 0.1  order_id  user_id eval_set  order_number  \
0              0   2539329        1    prior             1   
1              0   2539329        1    prior             1   
2              0   2539329        1    prior             1   
3              0   2539329        1    prior             1   
4              0   2539329        1    prior             1   
5              1   2398795        1    prior             2   
6              1   2398795        1    prior             2   
7              1   2398795        1    prior             2   
8              1   2398795        1    prior             2   
9              1   2398795        1    prior             2   
10             1   2398795        1    prior             2   
11             2    473747        1    prior             3   
12             2    473747        1    prior             3   
13             2    473747        1    prior             3   
14             2    473747        1    prior             3   
15             2    473747        1    prior             3   
16             3   2254736        1    prior             4   
17             3   2254736        1    prior             4   
18             3   2254736        1    prior             4   
19             3   2254736        1    prior             4   

    orders_days_of_week  order_hour_of_day  days_since_prior_order  \
0                     2                  8                     NaN   
1                     2                  8                     NaN   
2                     2                  8                     NaN   
3                     2                  8                     NaN   
4                     2                  8                     NaN   
5                     3                  7                    15.0   
6                     3                  7                    15.0   
7                     3                  7                    15.0   
8                     3                  7                    15.0   
9                     3                  7                    15.0   
10                    3                  7                    15.0   
11                    3                 12                    21.0   
12                    3                 12                    21.0   
13                    3                 12                    21.0   
14                    3                 12                    21.0   
15                    3                 12                    21.0   
16                    4                  7                    29.0   
17                    4                  7                    29.0   
18                    4                  7                    29.0   
19                    4                  7                    29.0   

    first_time_customers  product_id  ...  match    price_range_loc  \
0                   True         196  ...   both  Mid-range product   
1                   True       14084  ...   both  Mid-range product   
2                   True       12427  ...   both  Low-range product   
3                   True       26088  ...   both  Low-range product   
4                   True       26405  ...   both  Low-range product   
5                  False         196  ...   both  Mid-range product   
6                  False       10258  ...   both  Low-range product   
7                  False       12427  ...   both  Low-range product   
8                  False       13176  ...   both  Mid-range product   
9                  False       26088  ...   both  Low-range product   
10                 False       13032  ...   both  Low-range product   
11                 False         196  ...   both  Mid-range product   
12                 False       12427  ...   both  Low-range product   
13                 False       10258  ...   both  Low-range product   
14                 False       25133  ...   both  Mid-range product   
15                 False       30450  ...   both  Low-range product   
16                 Fals

In [18]:
# Create frequency_flag based on the criteria
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [19]:
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] > 10) & (ords_prods_merge['customer_frequency'] <=20), 'frequency_flag'] = 'Regular customer'

In [20]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [21]:
# Check the frequency for the frequency_flag
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [23]:
#Check the head for the new columns
ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(100)

user_id  customer_frequency         frequency_flag
0         1                20.5  Non-frequent customer
1         1                20.5  Non-frequent customer
2         1                20.5  Non-frequent customer
3         1                20.5  Non-frequent customer
4         1                20.5  Non-frequent customer
..      ...                 ...                    ...
95        2                13.0       Regular customer
96        2                13.0       Regular customer
97        2                13.0       Regular customer
98        2                13.0       Regular customer
99        2                13.0       Regular customer

[100 rows x 3 columns]

In [24]:
#Check shape 
ords_prods_merge.shape

(32404859, 27)

In [25]:
# Export dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merged_derived_new_variables_aggregated.pkl'))